src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,350 @@
+        return Q, R
+    else:
+        return Q
+
+
+    @defaults('p', 'q')
+    def random_generalized_svd(A, S=None, T=None, modes=None, p=20, q=2):
+    
+    """ Randomized SVD of a |VectorArray|. 
+    
+    Viewing the |VectorArray| 'A' as a 'A.dim' x 'len(A)' matrix, the return value
+    of this method is the randomized singular value decomposition of 'A', where the
+    inner product on R^('dim(A)') is given by 'S' and the inner product on R^('len(A)')
+    is given by 'T'. 
+        
+        A = U*\Sigma*V*T
+    
+    Parameters
+    ----------
+    A : 
+        The |VectorArray| or |Operator| for which the randomized SVD is to be computed. 
+    S : 
+        Range product |Operator| w.r.t which the randomized SVD is computed
+
+    T :
+        Source product |Operator| w.r.t which the randomized SVD is computed
+
+    modes : 
+        If not 'None', at most the first 'modes' approximated singular values 
+        and vectors are returned. 
+    p :
+        If not '0', adds 'Oversampling' colums to the randomly sampled matrix 'G'. 
+    q : 
+        If not '0', performs 'PowerIterations' iterations to increase the relative weight
+        of the first singular values.
+
+    Returns
+    -------
+    U 
+        |VectorArray| of approximated left singular vectors 
+    s 
+        One-dimensional |NumPy array| of the approximated singular values
+    V
+        |VectorArray| of the approximated right singular vectors
+    """
+    logger = getLogger('pymor.algorithms.rand_la')
+    
+    assert isinstance(A, VectorArray) or isinstance(A, Operator)
+    
+    if isinstance(A, VectorArray):
+        A = VectorArrayOperator(A)
+   
+    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes,int)
+    assert p >= 0 and isinstance(p, int)
+    assert q >= 0 and isinstance(q, int)
+    
+    if S is None:
+        S = IdentityOperator(A.range)
+    else:
+        assert isinstance(S, Operator)
+        
+    if T is None:
+        T = IdentityOperator(A.source)
+    else:
+        assert isinstance(T, Operator)
+        
+    if A.source.dim == 0 or A.range.dim == 0:
+        return A.source.empty(), np.array([]), A.range.empty()
+    
+    Q = rrf(A, source_product=T, range_product=S, q=q, l=modes+p)
+    B = A.apply_adjoint(S.apply(Q))
+    Q_B, R_B = gram_schmidt(T.apply_inverse(B), product=T, return_R=True)
+    U_b, s, V_b = sp.linalg.svd(R_B.T)
+
+    with logger.block(f'Computing generalized left-singular vectors ({U_b[:,:modes].shape[1]} vectors) ...'):
+        U = Q.lincomb(U_b)
+
+    with logger.block(f'Computung gerneralized right-singular vector ({V_b[:modes,:].shape[0]} vectors) ...'):
+        V = Q_B.lincomb(V_b.T)
+        
+    return  U[:modes], s[:modes], V[:modes]
+
+
+    @defaults('p', 'q','which','maxIter','complex_evp','left_evp','seed')
+    def random_ghep(A, E=None, k=None, p=20, q=2, sigma=None, which='LM', b=None, l=None, maxIter=1000, complex_evp=False, left_evp=False, seed=0):
```

Comment:
Many of the header arguments are not actually used in the function below, also `modes` is not part of the header.